### DISCUSSION TITLE
IncrementInteger and preg_match_all

### DISCUSSION THREAD

**Discussion by mfn** [infection/infection#1559] [Q&A] [answered]
Hi there,

infection ~n00b~ newcomer here 😄 

If an existing (much more complex) regex I use like this (very simplified example, no errors checks etc.):
```php
function foo($a) {
  if (preg_match_all("/(\d)/", $a, $m)) {
    return $m[0];
  }
  return [];
}

foo("1234");
```
var_dump of `$m` in the positive `if` branch:
```
array(2) {
  [0]=>
  array(4) {
    [0]=>
    string(1) "1"
    [1]=>
    string(1) "2"
    [2]=>
    string(1) "3"
    [3]=>
    string(1) "4"
  }
  [1]=>
  array(4) {
    [0]=>
    string(1) "1"
    [1]=>
    string(1) "2"
    [2]=>
    string(1) "3"
    [3]=>
    string(1) "4"
  }
}
```
My problem: the `IncrementInteger` mutator change the code to `$m[1];`:
```diff
---     return $m[0];
+++     return $m[1];
```

But how can I "fix" this, i.e I don't see what test could cover this.

Thanks! 🙏

**Comment by maks-rafalko** [comment, marked as answer]
Hi, on your example, `$m[0]` returns `1` while `$m[1]` returns `2`, so we can write a test that checks exact value.

If it doesn't matter which integer value is returned, then probably you can do the following:

1. replace `return $m[0]` with `current($m)`. `IncrementInteger` mutator in this case won't be applied
2. Ignore `IncrementInteger` mutator in this method or even line of code. See how: https://infection.github.io/guide/how-to.html#How-to-disable-Mutators-and-profiles

**Reply by mfn** [reply]
> Hi, on your example, `$m[0]` returns `1` while `$m[1]` returns `2`, so we can write a test that checks exact value.

Apologies I don't follow here 😅 `$m[0]` and `$m[1]` return the exact same array of 4 string (integer looking) values.
But…

> 2\. Ignore `IncrementInteger` mutator in this method or even line of code

Thanks, this is what I ended up doing for now!

I'm not sure I like to use `current()` but I think I understand what you're trying to see. If I use a method like `array_value_first` it would also be safe, unless there's a mutator turning it into a fictional `array_value_last` or so 😅

Thank you 🙏

**Comment by sanmai** [comment]
If you won't use the parentheses as in `"/\d/"` you won't have  `$m[1]` and the test would fail on this change.

**Reply by mfn** [reply]
Thanks, I'm ware of this behaviour.

In reality the regex is super complex and I don't dare to touch it; so I'll definitely stay with ignoring it

But thanks for pointing this out again, I'll remember this the next time and maybe can improve this then 👍
